### DISCUSSION TITLE
How to test Angular Clarity components with Jasmine

### DISCUSSION THREAD

**Discussion by beaker1977** [vmware-clarity/ng-clarity#783] [General]
I'm trying to write unit tests for my application and wantto validate whether various Clarity UI objects are or are not visible on the screen (eg. datagrids, vertical nav, text areas, selects, etc). I'm using Angular's Testbed functionality, however none of the code I've written is grabbing the UI elements so my test cases fail. How have others addressed this? The clarity-seed project has some sample test cases but they are just grabbing standard html elements (eg. <p>) but doesn't have any sample code trying to grab a datagrid for example.

**Comment by kevinbuhmann** [comment]
You can just use the `debugElement` like normal.

```typescript
const datagridDebugElement = fixture.debugElement.query(By.directive(ClrDatagrid));
const datagridInstance: ClrDatagrid = datagridDebugElement.componentInstance;
const datagridElement: HTMLElement = datagridDebugElement.nativeElement;
```

**Reply by beaker1977** [reply]
Thanks for the feedback. I've imported the ClairityModule in my .spec file but it doesn't recognize 'ClrDataGrid' reference. Is there some other directive I need to import?

import .......
import { ClarityModule } from "@clr/angular";

describe('LinksComponent', () => {
  let component: LinksComponent;
  let fixture: ComponentFixture<LinksComponent>;
  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [RouterTestingModule,
        HttpClientModule,
        ClarityModule
     ],
      providers:[
        LoggingService
      ],
      declarations: [LinksComponent]
    })
    .compileComponents();;
  });

  it('should create the component', () => {
    const fixture = TestBed.createComponent(LinksComponent);
    const app = fixture.componentInstance;
    expect(app).toBeDefined();
  });

  it('should render datagrid', () => {
    const datagridDebugElement = fixture.debugElement.query(By.directive(ClrDatagrid));
    const datagridInstance: ClrDatagrid = datagridDebugElement.componentInstance;
    const datagridElement: HTMLElement = datagridDebugElement.nativeElement
  });

**Reply by kevinbuhmann** [reply]
>I've imported the ClairityModule in my .spec file but it doesn't recognize 'ClrDataGrid' reference. Is there some other directive I need to import?

Well, you need to import `ClrDatagrid` from the `@clr/angular` package. You'll need to import all symbols that you use from their respective modules due to how JavaScript/TypeScript works.

**Reply by beaker1977** [reply]
So I imported the ClrDatagrid package (import { ClrDatagrid } from "@clr/angular";)and declared it as part of the TestBed config. However I get a 'Cannot read properties of undefined (reading 'debugElement')' error when my test script executes. Line 3 is throwing the error. Is something still missing in the spec file that it's not getting the datagrid component?

1 it('should render datagrid', () => {
2   const datagridDebugElement: DebugElement = fixture.debugElement.query(By.directive(ClrDatagrid));
3    const datagridInstance: ClrDatagrid = datagridDebugElement.componentInstance;
4   // const datagridElement: HTMLElement = datagridDebugElement.nativeElement
5   // expect(datagridElement).toBeDefined();
6  });


<div class="centeredContent" *ngIf="displayResults">
  <clr-datagrid>
      <clr-dg-column>Link</clr-dg-column>
      <clr-dg-column>Description</clr-dg-column>
      <clr-dg-column>Auth Method</clr-dg-column>

      <clr-dg-row *clrDgItems ="let link of links" [clrDgItem]="link">
          <clr-dg-cell><a href="{{link.link_url}}" target="_blank">{{link.link_nm}}</a></clr-dg-cell>
          <clr-dg-cell>{{link.link_desc}}</clr-dg-cell>
          <clr-dg-cell>{{link.link_auth}}</clr-dg-cell>
      </clr-dg-row>
  </clr-datagrid>
</div>

**Reply by kevinbuhmann** [reply]
>Cannot read properties of undefined (reading 'debugElement')

It's `fixture.debugElement`, so `fixture` must be `undefined`. That is a bug in your test code, then. You will need to assign `fixture`.

I am sorry, but this is not a general Angular help forum.

**Reply by beaker1977** [reply]
I found my bug and have finally been able to grab the datagrid and inspect it per my tests. I have another page that has two datagrids but the below code

const datagridDebugElement = fixture.debugElement.query(By.directive(ClrDatagrid));
const datagridInstance: ClrDatagrid = datagridDebugElement.componentInstance;
const datagridElement: HTMLElement = datagridDebugElement.nativeElement

only retrieves the first datagrid on the page. Is there a way to retrieve an 'array' of datagrids?

**Reply by kevinbuhmann** [reply]
Please read the Angular docs. https://angular.io/api/core/DebugElement
